File: controllers/gamepad/recoil_compensation.py
from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass, replace

from vision.recoil_collection.calibration import RecoilControlCalibration
from vision.recoil_collection.calibration import stick_from_pixels
from vision.recoil_collection.models import RecoilProfileRecord

from .state import GamepadFrame, GamepadOutput


_MANUAL_FIRE_TRIGGER_THRESHOLD = 10
_TARGET_DIRECTION_YIELD_SOFT_ERROR_PX = 8.0
_TARGET_DIRECTION_YIELD_FULL_ERROR_PX = 32.0
_TARGET_DIRECTION_YIELD_STALE_SECONDS = 0.080


@dataclass(slots=True, frozen=True)
class RecoilCompensationConfig:
    profile_amount: float = 1.0
    feedback_amount: float = 0.30
    profile_x_amount: float = 1.0
    profile_lead_ms: int = 0
    profile_velocity_reference_ms: int = 10
    profile_despike_enabled: bool = True
    profile_despike_threshold_px: float = 2.0
    profile_despike_ratio: float = 3.0
    target_direction_yield_enabled: bool = True
    selection_log_enabled: bool = True
    piecewise_mid_pixels_y: float = 45.0
    piecewise_max_pixels_y: float = 180.0
    piecewise_mid_ratio_y: float = 0.65


class RecoilCompensationPlugin:
    def __init__(
        self,
        config: RecoilCompensationConfig | None = None,
        *,
        profile_provider: Callable[
            [GamepadFrame],
            RecoilProfileRecord | tuple[RecoilProfileRecord | None, RecoilControlCalibration | None] | None,
        ]
        | None = None,
        profile_selection_logger: Callable[[str], None] | None = None,
    ):
        self.config = config or RecoilCompensationConfig()
        self._profile_provider = profile_provider
        self._profile_selection_logger = profile_selection_logger
        self._active_profile_id: str | None = None
        self._active_playback_profile: RecoilProfileRecord | None = None
        self._active_fire_started_at: float | None = None
        self._last_logged_selection_key: tuple[str, str | None] | None = None

    def reset(self) -> None:
        self._active_profile_id = None
        self._active_playback_profile = None
        self._active_fire_started_at = None

    def apply(self, frame: GamepadFrame, output: GamepadOutput) -> None:
        fire_active = output.auto_fire_active or _manual_fire_pressed(frame)
        if self._profile_provider is None:
            if fire_active and self.config.feedback_amount != 0.0:
                fallback_stick_y = -int(self.config.feedback_amount * 32767)
                output.right_y += int(round(_target_direction_yielded_stick(
                    fallback_stick_y,
                    frame=frame,
                    target_error_px=frame.target_dy,
                    desired_stick_direction=-frame.target_dy,
                    config=self.config,
                )))
            return

        profile, calibration = _coerce_profile_result(self._profile_provider(frame))
        if profile is None:
            self.reset()
            if fire_active:
                self._log_profile_selection(frame, None)
            if fire_active and self.config.feedback_amount != 0.0:
                fallback_stick_y = -int(self.config.feedback_amount * 32767)
                output.right_y += int(round(_target_direction_yielded_stick(
                    fallback_stick_y,
                    frame=frame,
                    target_error_px=frame.target_dy,
                    desired_stick_direction=-frame.target_dy,
                    config=self.config,
                )))
            return

        if self._active_profile_id != profile.profile_id:
            self.reset()
            self._active_profile_id = profile.profile_id
            self._active_playback_profile = _despiked_profile(profile, config=self.config)
        playback_profile = self._active_playback_profile or profile

        if not fire_active:
            self._active_fire_started_at = None
            return

        self._log_profile_selection(frame, profile)
        if self._active_fire_started_at is None:
            self._active_fire_started_at = float(frame.timestamp)

        elapsed_ms = max(0, int(round((float(frame.timestamp) - self._active_fire_started_at) * 1000.0)))
        profile_elapsed_ms = elapsed_ms + max(0, int(self.config.profile_lead_ms))
        cumulative_pixels_x, cumulative_pixels_y = _cumulative_profile_values(
            playback_profile,
            elapsed_ms=profile_elapsed_ms,
        )
        profile_scale = max(0.0, float(self.config.profile_amount))
        profile_x_scale = profile_scale * float(self.config.profile_x_amount)
        previous_pixels_x, previous_pixels_y = _cumulative_profile_values(
            playback_profile,
            elapsed_ms=max(0, profile_elapsed_ms - profile.sample_interval_ms),
        )
        delta_x = cumulative_pixels_x - previous_pixels_x
        delta_y = cumulative_pixels_y - previous_pixels_y
        if calibration is not None:
            anti_recoil_stick_y = (
                stick_from_pixels(
                    -delta_y,
                    axis="y",
                    duration_ms=profile.sample_interval_ms,
                    calibration=calibration,
                )
                * profile_scale
            )
            profile_stick_x = (
                stick_from_pixels(
                    -delta_x,
                    axis="x",
                    duration_ms=profile.sample_interval_ms,
                    calibration=calibration,
                )
                * profile_x_scale
            )
        else:
            mapped_delta_y = _normalize_delta_for_uncalibrated_mapping(
                delta_y,
                sample_interval_ms=profile.sample_interval_ms,
                config=self.config,
            )
            mapped_delta_x = _normalize_delta_for_uncalibrated_mapping(
                delta_x,
                sample_interval_ms=profile.sample_interval_ms,
                config=self.config,
            )
            anti_recoil_stick_y = (
                _map_pixels_to_stick(-mapped_delta_y, config=self.config)
                * profile_scale
            )
            profile_stick_x = _map_pixels_to_stick(-mapped_delta_x, config=self.config) * profile_x_scale
        profile_stick_x = _target_direction_yielded_stick(
            profile_stick_x,
            frame=frame,
            target_error_px=frame.target_dx,
            desired_stick_direction=frame.target_dx,
            config=self.config,
        )
        anti_recoil_stick_y = _target_direction_yielded_stick(
            anti_recoil_stick_y,
            frame=frame,
            target_error_px=frame.target_dy,
            desired_stick_direction=-frame.target_dy,
            config=self.config,
        )
        if profile_stick_x:
            output.right_x += int(round(profile_stick_x))
        if anti_recoil_stick_y:
            output.right_y += int(round(anti_recoil_stick_y))

    def _log_profile_selection(self, frame: GamepadFrame, profile: RecoilProfileRecord | None) -> None:
        if not self.config.selection_log_enabled:
            return
        logger = self._profile_selection_logger
        if logger is None:
            return

        aim_mode = "ads" if frame.is_aiming else "hipfire"
        profile_id = None if profile is None else profile.profile_id
        key = (aim_mode, profile_id)
        if self._last_logged_selection_key == key:
            return
        self._last_logged_selection_key = key

        if profile is None:
            logger(f"[Recoil] active_profile aim={aim_mode} profile=none fallback={self.config.feedback_amount:.0%}")
            return
        logger(
            f"[Recoil] active_profile aim={aim_mode} profile={profile.profile_id} "
            f"confidence={profile.confidence:.3f} profile_amount={self.config.profile_amount:.2f} "
            f"profile_x={self.config.profile_x_amount:.2f} feedback={self.config.feedback_amount:.2f} "
            f"lead_ms={max(0, int(self.config.profile_lead_ms))} "
            f"velocity_ref_ms={max(1, int(self.config.profile_velocity_reference_ms))}"
        )


def _coerce_profile_result(
    value: RecoilProfileRecord | tuple[RecoilProfileRecord | None, RecoilControlCalibration | None] | None,
) -> tuple[RecoilProfileRecord | None, RecoilControlCalibration | None]:
    if value is None:
        return None, None
    if isinstance(value, tuple) and len(value) == 2:
        return value[0], value[1]
    return value, None


def _cumulative_profile_values(profile: RecoilProfileRecord, *, elapsed_ms: int) -> tuple[float, float]:
    if elapsed_ms < profile.initial_delay_ms:
        return 0.0, 0.0

    sample_index = (elapsed_ms - profile.initial_delay_ms) // profile.sample_interval_ms
    sample_index = max(0, min(sample_index, profile.sample_count - 1))
    return float(profile.samples_x[sample_index]), float(profile.samples_y[sample_index])


def _despiked_profile(profile: RecoilProfileRecord, *, config: RecoilCompensationConfig) -> RecoilProfileRecord:
    if not config.profile_despike_enabled or profile.sample_count < 3:
        return profile

    samples_x = _despike_cumulative_samples(
        profile.samples_x,
        threshold_px=float(config.profile_despike_threshold_px),
        ratio=float(config.profile_despike_ratio),
    )
    samples_y = _despike_cumulative_samples(
        profile.samples_y,
        threshold_px=float(config.profile_despike_threshold_px),
        ratio=float(config.profile_despike_ratio),
    )
    if samples_x == profile.samples_x and samples_y == profile.samples_y:
        return profile
    return replace(profile, samples_x=samples_x, samples_y=samples_y)


def _despike_cumulative_samples(
    values: tuple[float, ...],
    *,
    threshold_px: float,
    ratio: float,
) -> tuple[float, ...]:
    if len(values) < 3:
        return values
    threshold = max(0.0, float(threshold_px))
    ratio = max(1.0, float(ratio))
    repaired = list(float(value) for value in values)
    original = tuple(float(value) for value in values)
    for index in range(1, len(original) - 1):
        prev_value = original[index - 1]
        current = original[index]
        next_value = original[index + 1]
        prev_delta = current - prev_value
        next_delta = next_value - current
        if prev_delta == 0.0 or next_delta == 0.0:
            continue
        if (prev_delta > 0.0) == (next_delta > 0.0):
            continue
        local_midpoint = (prev_value + next_value) * 0.5
        spike_error = current - local_midpoint
        neighbor_span = abs(next_value - prev_value)
        if abs(spike_error) < threshold:
            continue
        if abs(spike_error) < ratio * max(threshold, neighbor_span):
            continue
        repaired[index] = local_midpoint
    return tuple(repaired)


def _manual_fire_pressed(frame: GamepadFrame) -> bool:
    return bool(frame.buttons.get("rb", False) or frame.right_trigger > _MANUAL_FIRE_TRIGGER_THRESHOLD)


def _target_direction_yielded_stick(
    recoil_stick: float,
    *,
    frame: GamepadFrame,
    target_error_px: float,
    desired_stick_direction: float,
    config: RecoilCompensationConfig,
) -> float:
    if not config.target_direction_yield_enabled:
        return recoil_stick
    if recoil_stick == 0.0 or desired_stick_direction == 0.0:
        return recoil_stick
    if not _has_fresh_target_direction(frame):
        return recoil_stick
    if (recoil_stick > 0.0) == (desired_stick_direction > 0.0):
        return recoil_stick

    error_px = abs(float(target_error_px))
    if error_px <= _TARGET_DIRECTION_YIELD_SOFT_ERROR_PX:
        return recoil_stick
    if error_px >= _TARGET_DIRECTION_YIELD_FULL_ERROR_PX:
        return 0.0

    yield_window = _TARGET_DIRECTION_YIELD_FULL_ERROR_PX - _TARGET_DIRECTION_YIELD_SOFT_ERROR_PX
    keep_scale = (_TARGET_DIRECTION_YIELD_FULL_ERROR_PX - error_px) / yield_window
    return recoil_stick * max(0.0, min(1.0, keep_scale))


def _has_fresh_target_direction(frame: GamepadFrame) -> bool:
    if frame.target_timestamp is None:
        return False
    age_seconds = float(frame.timestamp) - float(frame.target_timestamp)
    return 0.0 <= age_seconds <= _TARGET_DIRECTION_YIELD_STALE_SECONDS


def _map_pixels_to_stick(delta: float, *, config: RecoilCompensationConfig) -> float:
    abs_delta = abs(float(delta))
    sign = -1.0 if delta < 0.0 else 1.0
    mid_pixels = float(config.piecewise_mid_pixels_y)
    max_pixels = float(config.piecewise_max_pixels_y)
    mid_ratio = float(config.piecewise_mid_ratio_y)

    if mid_pixels <= 0.0 or max_pixels <= mid_pixels or mid_ratio <= 0.0 or mid_ratio >= 1.0:
        clamped = max(-max_pixels, min(max_pixels, float(delta)))
        return (clamped / max_pixels) * 32767.0 if max_pixels > 0.0 else 0.0

    if abs_delta >= max_pixels:
        return sign * 32767.0
    if abs_delta <= mid_pixels:
        return sign * 32767.0 * mid_ratio * (abs_delta / mid_pixels)

    progress = (abs_delta - mid_pixels) / (max_pixels - mid_pixels)
    return sign * 32767.0 * (mid_ratio + ((1.0 - mid_ratio) * progress))


def _normalize_delta_for_uncalibrated_mapping(
    delta: float,
    *,
    sample_interval_ms: int,
    config: RecoilCompensationConfig,
) -> float:
    interval_ms = max(1, int(sample_interval_ms))
    reference_ms = max(1, int(config.profile_velocity_reference_ms))
    return float(delta) * (reference_ms / interval_ms)

File: native/controller_native/recoil_compensation.cpp
#include "recoil_compensation.h"

#include <algorithm>
#include <cmath>
#include <iomanip>
#include <iostream>
#include <sstream>
#include <utility>

namespace controller_native {

namespace {

float clamp_unit(float value) {
    return std::max(-1.0f, std::min(1.0f, value));
}

float fallback_amount(const GamepadRecoilConfig& config) {
    const float minimum = std::max(0.0f, config.feedback_min_amount);
    const float maximum = std::max(minimum, config.feedback_max_amount);
    return std::clamp(config.feedback_amount, minimum, maximum);
}

}  // namespace

NativeRecoilCompensation::NativeRecoilCompensation(GamepadRecoilConfig config)
    : config_(std::move(config)),
      profile_directory_(config_.profile_directory),
      calibration_directory_(config_.calibration_directory),
      recognizer_state_path_(config_.recognizer_state_path) {}

void NativeRecoilCompensation::reset() {
    fire_started_at_seconds_ = 0.0;
}

bool NativeRecoilCompensation::load_profile_directory(const std::filesystem::path& directory) {
    profile_directory_ = directory;
    const std::optional<RecoilProfile> profile = load_first_recoil_profile(directory);
    if (!profile.has_value() || profile->empty()) {
        return false;
    }
    set_profile(*profile);
    return true;
}

bool NativeRecoilCompensation::load_calibration_directory(const std::filesystem::path& directory) {
    calibration_directory_ = directory;
    if (!active_profile_.has_value()) {
        active_calibration_.reset();
        return std::filesystem::exists(directory);
    }
    active_calibration_ = load_active_calibration(*active_profile_);
    return active_calibration_.has_value();
}

void NativeRecoilCompensation::set_recognizer_state_path(
    const std::filesystem::path& recognizer_state_path) {
    recognizer_state_path_ = recognizer_state_path;
}

void NativeRecoilCompensation::set_profile(const RecoilProfile& profile) {
    active_profile_ = profile;
    playback_profile_ = build_recoil_playback_profile(profile, config_);
    active_calibration_ = load_active_calibration(profile);
    reset();
}

NativeRecoilOutput NativeRecoilCompensation::compute(const NativeRecoilInput& input) {
    NativeRecoilOutput output;
    if (!config_.enabled) {
        return output;
    }
    if (!input.fire_active) {
        reset();
        return output;
    }

    output.recoil_active = true;
    if (!config_.profile_playback_enabled) {
        output.right_y_delta = -fallback_amount(config_);
        return output;
    }

    select_runtime_profile_for_context(input.aiming);
    if (!playback_profile_.has_value() || playback_profile_->empty()) {
        output.right_y_delta = -fallback_amount(config_);
        return output;
    }

    if (fire_started_at_seconds_ <= 0.0) {
        fire_started_at_seconds_ = input.now_seconds;
    }

    // timeline: advance continuously while fire is active; do not pause on target-direction yield.
    const int elapsed_ms = std::max(
        0,
        static_cast<int>(std::lround((input.now_seconds - fire_started_at_seconds_) * 1000.0)));
    const int profile_elapsed_ms =
        elapsed_ms + std::max(0, static_cast<int>(std::lround(config_.profile_lead_ms)));
    const auto [cumulative_x, cumulative_y] = cumulative_profile_values(profile_elapsed_ms);
    const auto [previous_x, previous_y] = cumulative_profile_values(
        std::max(0, profile_elapsed_ms - playback_profile_->sample_interval_ms));
    const float delta_x = cumulative_x - previous_x;
    const float delta_y = cumulative_y - previous_y;

    const float profile_scale = std::max(0.0f, config_.profile_amount);
    const float profile_x_scale = profile_scale * config_.profile_x_amount;
    float profile_stick_x = 0.0f;
    float anti_recoil_stick_y = 0.0f;
    if (active_calibration_.has_value()) {
        profile_stick_x = map_calibrated_pixels_to_stick(
            -delta_x,
            "x",
            playback_profile_->sample_interval_ms) * profile_x_scale;
        anti_recoil_stick_y = map_calibrated_pixels_to_stick(
            -delta_y,
            "y",
            playback_profile_->sample_interval_ms) * profile_scale;
    } else {
        const float mapped_delta_x = normalize_delta_for_uncalibrated_mapping(
            delta_x,
            playback_profile_->sample_interval_ms);
        const float mapped_delta_y = normalize_delta_for_uncalibrated_mapping(
            delta_y,
            playback_profile_->sample_interval_ms);
        profile_stick_x = map_pixels_to_stick(-mapped_delta_x) * profile_x_scale;
        anti_recoil_stick_y = map_pixels_to_stick(-mapped_delta_y) * profile_scale;
    }

    // profile_despike is applied by build_recoil_playback_profile; raw JSON stays untouched.
    output.right_x_delta = profile_stick_x;
    output.right_y_delta = anti_recoil_stick_y;
    return output;
}

void NativeRecoilCompensation::select_runtime_profile_for_context(bool aiming) {
    if (profile_directory_.empty() || recognizer_state_path_.empty()) {
        return;
    }
    const std::string aim_mode = aiming ? "ads" : "hipfire";
    const std::optional<RecoilProfile> selected =
        load_matching_recoil_profile(profile_directory_, recognizer_state_path_, aim_mode);
    if (!selected.has_value() || selected->empty()) {
        active_profile_.reset();
        playback_profile_.reset();
        active_calibration_.reset();
        active_aim_mode_.clear();
        reset();
        log_profile_selection(aim_mode, nullptr);
        return;
    }
    if (active_profile_.has_value() &&
        active_profile_->profile_id == selected->profile_id &&
        active_aim_mode_ == aim_mode) {
        log_profile_selection(aim_mode, &(*active_profile_));
        return;
    }
    set_profile(*selected);
    active_aim_mode_ = aim_mode;
    log_profile_selection(aim_mode, &(*active_profile_));
}

void NativeRecoilCompensation::log_profile_selection(
    const std::string& aim_mode,
    const RecoilProfile* profile) {
    if (!config_.selection_log_enabled) {
        return;
    }
    const std::string profile_id = profile == nullptr ? "none" : profile->profile_id;
    if (has_logged_selection_ &&
        last_logged_aim_mode_ == aim_mode &&
        last_logged_profile_id_ == profile_id) {
        return;
    }
    has_logged_selection_ = true;
    last_logged_aim_mode_ = aim_mode;
    last_logged_profile_id_ = profile_id;

    std::ostringstream line;
    if (profile == nullptr) {
        const int fallback_percent = static_cast<int>(
            std::lround(std::max(0.0f, config_.feedback_amount) * 100.0f));
        line
            << "[Recoil] active_profile aim=" << aim_mode
            << " profile=none fallback=" << fallback_percent << "%";
    } else {
        line
            << std::fixed
            << "[Recoil] active_profile aim=" << aim_mode
            << " profile=" << profile->profile_id
            << " confidence=" << std::setprecision(3) << profile->confidence
            << " profile_amount=" << std::setprecision(2) << config_.profile_amount
            << " profile_x=" << std::setprecision(2) << config_.profile_x_amount
            << " feedback=" << std::setprecision(2) << config_.feedback_amount
            << " lead_ms=" << std::max(0, static_cast<int>(std::lround(config_.profile_lead_ms)))
            << " velocity_ref_ms="
            << std::max(1, static_cast<int>(std::lround(config_.profile_velocity_reference_ms)));
    }
    std::cout << line.str() << '\n';
}

std::pair<float, float> NativeRecoilCompensation::cumulative_profile_values(int elapsed_ms) const {
    if (!playback_profile_.has_value() || playback_profile_->empty() ||
        elapsed_ms < playback_profile_->initial_delay_ms) {
        return {0.0f, 0.0f};
    }
    const int sample_index = std::max(
        0,
        std::min(
            static_cast<int>(playback_profile_->sample_count()) - 1,
            (elapsed_ms - playback_profile_->initial_delay_ms) /
                std::max(1, playback_profile_->sample_interval_ms)));
    return {
        playback_profile_->samples_x[static_cast<std::size_t>(sample_index)],
        playback_profile_->samples_y[static_cast<std::size_t>(sample_index)],
    };
}

std::optional<RecoilCalibration> NativeRecoilCompensation::load_active_calibration(
    const RecoilProfile& profile) const {
    return load_matching_recoil_calibration(calibration_directory_, profile);
}

float NativeRecoilCompensation::map_calibrated_pixels_to_stick(
    float pixels,
    const std::string& axis,
    int duration_ms) const {
    if (!active_calibration_.has_value()) {
        return 0.0f;
    }
    return calibrated_stick_from_pixels(pixels, axis, duration_ms, *active_calibration_);
}

float NativeRecoilCompensation::normalize_delta_for_uncalibrated_mapping(
    float delta,
    int sample_interval_ms) const {
    const int interval_ms = std::max(1, sample_interval_ms);
    const int reference_ms = std::max(
        1,
        static_cast<int>(std::lround(config_.profile_velocity_reference_ms)));
    return delta * (static_cast<float>(reference_ms) / static_cast<float>(interval_ms));
}

float NativeRecoilCompensation::map_pixels_to_stick(float delta) const {
    const float abs_delta = std::fabs(delta);
    const float sign = delta < 0.0f ? -1.0f : 1.0f;
    const float mid_pixels = config_.piecewise_mid_pixels_y;
    const float max_pixels = config_.piecewise_max_pixels_y;
    const float mid_ratio = config_.piecewise_mid_ratio_y;
    if (mid_pixels <= 0.0f || max_pixels <= mid_pixels || mid_ratio <= 0.0f || mid_ratio >= 1.0f) {
        return max_pixels > 0.0f ? clamp_unit(delta / max_pixels) : 0.0f;
    }
    if (abs_delta >= max_pixels) {
        return sign;
    }
    if (abs_delta <= mid_pixels) {
        return sign * mid_ratio * (abs_delta / mid_pixels);
    }
    const float progress = (abs_delta - mid_pixels) / (max_pixels - mid_pixels);
    return sign * (mid_ratio + ((1.0f - mid_ratio) * progress));
}

}  // namespace controller_native
